File: BST/lua/tree.lua
require "node"

Tree = {}

function Tree.new()
    local tree = {root = nil}
    tree.add = Tree.add
    tree.contain = Tree.contain
    tree.length = Tree.length
    tree.height = Tree.height
    return tree
end

function Tree:add(value)
    if self.root == nil then
        self.root = Node.new(value)
        return true
    end
    return self.root:add(value)
   
end

function Tree:contain(value)
    if self.root == nil then
        return false
    end
    return self.root:contain(value)
end

function Tree:length(value)
    if self.root == nil then
        return 0
    end
    return self.root:length(value)
end

function Tree:height(value)
    if self.root == nil then
        return 0
    end
    return self.root:height(value)
end

File: BST/java/Tree.java
class Tree{
    Node root;

    public Tree(){
        this.root = null;
    }

    public boolean add(int value){
        if(this.root == null){
            this.root = new Node(value);
            return true;
        }
        return this.root.add(value);
    }

    public boolean contain(int value){        
        if(this.root == null){
            return false;
        }
        return this.root.contain(value);
    }

    public int length(){
        if(this.root == null){
            return 0;
        }
        return this.root.length();
    }
    public int height(){
        if(this.root == null){
            return 0;
        }
        return this.root.height();
    }
}